executor/insert_common.go

```diff
@@ -792,6 +807,62 @@ func getAutoRecordID(d types.Datum, target *types.FieldType, isInsert bool) (int
 	return recordID, nil
 }
 
+func (e *InsertValues) adjustAutoShardDatum(ctx context.Context, d types.Datum, hasValue bool, c *table.Column) (types.Datum, error) {
+	if !hasValue || d.IsNull() {
+		autoShardID, err := e.allocAutoShardID(&c.FieldType)
+		if err != nil {
+			return types.Datum{}, err
+		}
+		d.SetAutoID(autoShardID, c.Flag)
+	} else {
+		recordID, err := getAutoRecordID(d, &c.FieldType, true)
+		if err != nil {
+			return types.Datum{}, err
+		}
+		err = e.rebaseAutoShardID(recordID, &c.FieldType)
+		if err != nil {
+			return types.Datum{}, err
+		}
+		d.SetAutoID(recordID, c.Flag)
+	}
+
+	casted, err := table.CastValue(e.ctx, d, c.ToInfo())
+	if err != nil {
+		return types.Datum{}, err
+	}
+	return casted, nil
+}
+
+// AllocAutoShardID allocates a shard id for primary key column. It assumes the tableInfo.AutoShardBits is greater than zero.
+func (e *InsertValues) allocAutoShardID(fieldType *types.FieldType) (int64, error) {
```

Comment:
better move these two to table.go

Reply:
Moving these functions to table/tables.go means we need to change the interface `Table`, which contains implementations like `infoschemaTable`, `perfSchemaTable` and `VirtualTable`. However, `auto_shard` does not apply to these implementations, it can bring a lot of dead code and violate the intuition.